corehq/apps/sso/models.py

```diff
@@ -1 +1,105 @@
 from django.db import models
+from django.contrib.postgres.fields import ArrayField
+
+from corehq.apps.accounting.models import BillingAccount
+
+
+class IdentityProviderType(object):
+    AZURE_AD = 'azure_ad'
+    CHOICES = (
+        (AZURE_AD, "Azure AD"),
+    )
+
+
+class IdentityProvider(models.Model):
+    """
+    This stores the information necessary to make a SAML request to an external
+    IdP. Right now this process supports Azure AD and the plan is to add
+    support for other identity provider types in the future.
+    """
+    # these three fields must only ever be editable by Accounting admins
+    name = models.CharField(max_length=128)
+    slug = models.CharField(max_length=256, db_index=True, unique=True)
+    idp_type = models.CharField(
+        max_length=10,
+        default=IdentityProviderType.AZURE_AD,
+        choices=IdentityProviderType.CHOICES,
+    )
+
+    # whether an IdP is editable by its BillingAccount owner
+    # (it will always be editable by accounting admins)
+    is_editable = models.BooleanField(default=False)
+
+    # whether an IdP is actively in use as an authentication method on HQ
+    is_active = models.BooleanField(default=False)
+
+    # the enterprise admins of this account will be able to edit the SAML
+    # configuration fields
+    owner = models.ForeignKey(BillingAccount, on_delete=models.PROTECT)
+
+    # these are fields required by the external IdP to form a SAML request
+    entity_id = models.TextField(blank=True, null=True)
+    login_url = models.TextField(blank=True, null=True)
+    logout_url = models.TextField(blank=True, null=True)
+    idp_cert_public = models.TextField(blank=True, null=True)
+
+    # the date the IdP's SAML signing certificate expires.
+    # this will be filled out by enterprise admins
+    date_idp_cert_expiration = models.DateTimeField(blank=True, null=True)
+
+    # as the service provider, this will store our x509 certificates and
+    # will be renewed automatically by a periodic task
+    sp_cert_public = models.TextField(blank=True, null=True)
```

Comment:
Thanks for that link. That was a good read. I initially approached making sure the SP assertions were also signed because in most of the SAML security articles I came across, that was always mentioned as an area of vulnerability and I wanted to make this setup as secure as possible.

I generated separate signing keys per IdP in the event one gets compromised it doesn't open up other assertions to being sniffed. The TLS point makes some sense though. 

> These requests can also be signed but in practice mostly breaks things when the SP rotates their key. There’s little security benefit since the SAML exchange is all happening over TLS in any modern system.

The plan is to use the rollover cert feature of `python3-saml` that should work around this. Is there any harm in keeping this in place for a super extra layer of security?